File: Algorithms/src/2127. Maximum Employees to Be Invited to a Meeting/Code/solution.go
func maximumInvitations(favorite []int) int {
    n := len(favorite)
    inDegree := make([]int, n)
    chainLengths := make([]int, n)
    visited := make([]bool, n)

    for _, fav := range favorite {
        inDegree[fav]++
    }

    queue := []int{}
    for i := 0; i < n; i++ {
        if inDegree[i] == 0 {
            queue = append(queue, i)
        }
    }

    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]
        visited[node] = true

        next := favorite[node]
        chainLengths[next] = chainLengths[node] + 1
        inDegree[next]--
        if inDegree[next] == 0 {
            queue = append(queue, next)
        }
    }

    maxCycle, totalChains := 0, 0
    for i := 0; i < n; i++ {
        if !visited[i] {
            current, cycleLength := i, 0
            for !visited[current] {
                visited[current] = true
                current = favorite[current]
                cycleLength++
            }

            if cycleLength == 2 {
                totalChains += 2 + chainLengths[i] + chainLengths[favorite[i]]
            } else {
                if cycleLength > maxCycle {
                    maxCycle = cycleLength
                }
            }
        }
    }

    if maxCycle > totalChains {
        return maxCycle
    }
    return totalChains
}

File: Algorithms/src/2127. Maximum Employees to Be Invited to a Meeting/Code/solution.py
class Solution:
    def maximumInvitations(self, favorite: List[int]) -> int:
        n = len(favorite)
        inDegree = [0] * n
        chainLengths = [0] * n
        visited = [False] * n

        for fav in favorite:
            inDegree[fav] += 1

        queue = deque(i for i in range(n) if inDegree[i] == 0)
        while queue:
            node = queue.popleft()
            visited[node] = True

            next_node = favorite[node]
            chainLengths[next_node] = chainLengths[node] + 1
            inDegree[next_node] -= 1
            if inDegree[next_node] == 0:
                queue.append(next_node)

        maxCycle = 0
        totalChains = 0
        for i in range(n):
            if not visited[i]:
                current, cycleLength = i, 0
                while not visited[current]:
                    visited[current] = True
                    current = favorite[current]
                    cycleLength += 1

                if cycleLength == 2:
                    totalChains += 2 + chainLengths[i] + chainLengths[favorite[i]]
                else:
                    maxCycle = max(maxCycle, cycleLength)

        return max(maxCycle, totalChains)
